### PR TITLE
openstack: remove the service VM

### PR DESCRIPTION
**Pull request by tomassedovic**
openstack: Remove the Service VM

The experimental OpenStack backend used to create an extra server
running DNS and load balancer services that the cluster needed.
OpenStack does not always come with DNSaaS or LBaaS so we had to provide
the functionality the OpenShift cluster depends on (e.g. the etcd SRV
records, the api-int records & load balancing, etc.).

This approach is undesirable for two reasons: first, it adds an extra
node that the other IPI platforms do not need. Second, this node is a
single point of failure.

The Baremetal platform has faced the same issues and they have solved
them with a few virtual IP addresses managed by keepalived in
combination with coredns static pod running on every node using the mDNS
protocol to update records as new nodes are added or removed and a
similar static pod haproxy to load balance the control plane internally.

The VIPs are defined here in the installer and they use the
PlatformStatus field to be passed to the necessary
machine-config-operator fields:

https://github.com/openshift/api/pull/374

The Bare Metal IPI Networking Infrastructure document is broadly
applicable here as well:

https://github.com/openshift/installer/blob/master/docs/design/baremetal/networking-infrastructure.md

Notable differences in OpenStack:

* We only use the API and DNS VIPs right now
* Instead of Baremetal's Ingress VIP (which is attached to the OpenShift
  routers) our haproxy static pods balance the 80 & 443 pods to the
  worker nodes
* We do not run coredns on the bootstrap node. Instead, bootstrap itself
  uses one of the masters for DNS.

These differences are not fundamental to OpenStack and we will be
looking at aligning more closely with the Baremetal provider in the
future.

There is also a great oportunity to share some of the configuration
files and scripts here.

This change needs several other pull requests:

Keepalived plus the coredns & haproxy static pods in the MCO:
openshift/machine-config-operator/pull/740

Passing the API and DNS VIPs through the installer:
https://github.com/openshift/installer/pull/1998

Vendoring the OpenStack PlatformStatus changes in the MCO:
https://github.com/openshift/machine-config-operator/pull/978

Allowing to use PlatformStatus in the MCO templates:
https://github.com/openshift/machine-config-operator/pull/943

Co-authored-by: Emilio Garcia <egarcia@redhat.com>
Co-authored-by: John Trowbridge <trown@redhat.com>
Co-authored-by: Martin Andre <m.andre@redhat.com>
Co-authored-by: Tomas Sedovic <tsedovic@redhat.com>

Massive thanks to the Bare Metal and oVirt people!

### CODE REVIEW

**Comment by tomassedovic** [comment]
Right now, I'm just pushing this out there so we can discuss & collaborate on it in public.

Most of the code has been written by @trown, @mandre and @iamemilio. I consolidated it and rebased on master.

**Comment by tomassedovic** [comment]
Note regarding the e2e-openstack job: it will keep failing until https://github.com/openshift/machine-config-operator/pull/740 is merged.

File: pkg/types/openstack/platform.go [hunk 1]
```diff
@@ -28,6 +28,14 @@ type Platform struct {
 	// Existing Floating IP to associate with the OpenStack load balancer.
 	LbFloatingIP string `json:"lbFloatingIP"`
 
+	// APIVIP
```
Comment (abhinavdahiya) [inline pkg/types/openstack/platform.go]: this needs better doc comments.

File: pkg/types/openstack/platform.go [hunk 2]
```diff
@@ -28,6 +28,14 @@ type Platform struct {
...
+	// IP in the machineCIDR to use for api-int.
+	APIVIP string `json:"apiVIP"`
+
+	// DNSVIP
```
Comment (abhinavdahiya) [inline pkg/types/openstack/platform.go]: this needs better doc commetns.

File: pkg/types/openstack/defaults/platform.go [hunk 3]
```diff
@@ -1,9 +1,37 @@
...
+
+	"github.com/apparentlymart/go-cidr/cidr"
+
+	"github.com/openshift/installer/pkg/types"
 	"github.com/openshift/installer/pkg/types/openstack"
 )
 
+// Defaults for the openstack platform.
+const (
+	APIVIP = ""
```
Comment (abhinavdahiya) [inline pkg/types/openstack/defaults/platform.go]: why they public? If they need to be public this needs comments.

File: pkg/types/openstack/defaults/platform.go [hunk 4]
```diff
@@ -1,9 +1,37 @@
...
+	DNSVIP = ""
+)
+
 // SetPlatformDefaults sets the defaults for the platform.
-func SetPlatformDefaults(p *openstack.Platform) {
+func SetPlatformDefaults(p *openstack.Platform, c *types.InstallConfig) {
```
Comment (abhinavdahiya) [inline pkg/types/openstack/defaults/platform.go]: installconfig has the platform.. i think it's better to pass the networking.

File: data/data/openstack/topology/private-network.tf [hunk 5]
```diff
@@ -78,28 +77,43 @@ resource "openstack_networking_port_v2" "bootstrap_port" {
...
 
-resource "openstack_networking_floatingip_associate_v2" "service_fip" {
+// Assign the floating IP to one of the masters.
+//
+// Strictly speaking, this is not required to finish the installation. We
+// support environments without floating IPs. However, since the installer
+// is running outside of the nodes subnet (often outside of the OpenStack
+// cluster itself), it needs a floating IP to monitor the progress.
+//
+// This IP address is not expected to be the final solution for providing HA.
```
Comment (abhinavdahiya) [inline data/data/openstack/topology/private-network.tf]: cc @crawford

Comment (tomassedovic) [reply data/data/openstack/topology/private-network.tf]: Alternatively, we could just require floating IP support for this first iteration and make this non-optional.

In general, VMs without a floating IP attached are not accessible from the outside (their IP addresses are private). So without a floating IP, the installer or anyone else can't reach the OpenShift API running there.

Some OpenStack environments do not allow any floating IPs at all or limit the amount the user can get.

File: pkg/types/openstack/defaults/platform.go (hunk 3 above)
Comment (tomassedovic) [reply pkg/types/openstack/defaults/platform.go]: Oops, these are leftovers from previous iterations. They're not actually used anywhere. Going to remove them.

File: pkg/types/openstack/defaults/platform.go (hunk 4 above)
Comment (tomassedovic) [reply pkg/types/openstack/defaults/platform.go]: We just did the same thing baremetal and vsphere did:

https://github.com/openshift/installer/blob/fc8c68319948411e384005a2ca55266c5fd0e200/pkg/types/baremetal/defaults/platform.go#L23

https://github.com/openshift/installer/blob/fc8c68319948411e384005a2ca55266c5fd0e200/pkg/types/defaults/installconfig.go#L72-L74

But I'll pass `Networking` in directly.

File: pkg/types/openstack/platform.go (hunk 1 above)
Comment (tomassedovic) [reply pkg/types/openstack/platform.go]: Done

**Comment by tomassedovic** [comment]
@abhinavdahiya thanks for your comments! They should all be addressed now.

Please let me know if there's anything else.

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template [hunk 6]
```diff
@@ -211,13 +211,17 @@ then
...
+	mkdir --parents /etc/mcc/bootstrap /etc/mcs/bootstrap /etc/kubernetes/manifests /etc/kubernetes/static-pod-resources
 	cp mco-bootstrap/bootstrap/manifests/* /etc/mcc/bootstrap/
 	cp openshift/* /etc/mcc/bootstrap/
 	# 4. read ImageContentSourcePolicy objects generated by the installer
 	cp manifests/* /etc/mcc/bootstrap/
 	cp auth/kubeconfig-kubelet /etc/mcs/kubeconfig
 	cp mco-bootstrap/bootstrap/machineconfigoperator-bootstrap-pod.yaml /etc/kubernetes/manifests/
+	if [ -d mco-bootstrap/openstack/manifests ]; then
+	    cp mco-bootstrap/openstack/manifests/* /etc/kubernetes/manifests/
+	    cp -r mco-bootstrap/openstack/static-pod-resources/* /etc/kubernetes/static-pod-resources/
```
Comment (abhinavdahiya) [inline data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: nit: I think this script uses full names for flags.. `--recursive`

File: pkg/types/openstack/platform.go [hunk 7]
```diff
@@ -28,6 +28,33 @@ type Platform struct {
 	// Existing Floating IP to associate with the OpenStack load balancer.
 	LbFloatingIP string `json:"lbFloatingIP"`
 
+	// APIVIP
+	// The internal virtual IP address (VIP) put in front of the
+	// Kubernetes API server for use by components inside the cluster.
+	// The DNS static pods running on the nodes resolve the api-int
+	// record to APIVIP.
+	//
+	// The value is set by the installer from the MachineCIDR range.
```
Comment (abhinavdahiya) [inline pkg/types/openstack/platform.go]: Can the users change this value?? If yes is there validation that it belongs to the machinecidr...?

Comment (tomassedovic) [reply pkg/types/openstack/platform.go]: Good catch, thanks!

The IPI users are not supposed to set this, but it is possible to do so in `install-config.yaml`. In that case there is no validation there. I'm going to add it.

File: data/data/bootstrap/files/usr/local/bin/bootkube.sh.template (hunk 6 above)
Comment (tomassedovic) [reply data/data/bootstrap/files/usr/local/bin/bootkube.sh.template]: We just used what baremetal did, but I agree being explicit is better. Fixed.

**Comment by tomassedovic** [comment]
/hold

Going to add validation for the VIP values.

**Comment by tomassedovic** [comment]
@abhinavdahiya would something like this be a good validation of the OpenStack VIPs?

https://github.com/tomassedovic/installer/commit/a8153b6e990851be43d5983ba6e40bdbfcaddfbd

I've pushed it to a new branch in case it breaks something. It works manually as well as with go tests, but I'm not super familiar with the testing here so there may be a better way to do this.

I'm not happy about passing the `Networking` struct to the platform tests for example, but I'm not sure if there's a better way to do this.

File: pkg/types/openstack/platform.go (hunk 7 above)
Comment (abhinavdahiya) [reply pkg/types/openstack/platform.go]: > The IPI users are not supposed to set this

will anything fail if they do??

**Comment by abhinavdahiya** [comment]
> @abhinavdahiya would something like this be a good validation of the OpenStack VIPs?
> 
> [tomassedovic@a8153b6](https://github.com/tomassedovic/installer/commit/a8153b6e990851be43d5983ba6e40bdbfcaddfbd)
> 
> I've pushed it to a new branch in case it breaks something. It works manually as well as with go tests, but I'm not super familiar with the testing here so there may be a better way to do this.
> 
> I'm not happy about passing the `Networking` struct to the platform tests for example, but I'm not sure if there's a better way to do this.

looks reasonable.

**Comment by tomassedovic** [comment]
Thanks! It's EOD here, I'll finish it up tomorrow.

File: pkg/types/openstack/platform.go (hunk 7 above)
Comment (tomassedovic) [reply pkg/types/openstack/platform.go]: It didn't before but it will now.

**Comment by tomassedovic** [comment]
@abhinavdahiya the VIP validation has been added.

I'll keep the `/hold` until the corresponding MCO merges: https://github.com/openshift/machine-config-operator/pull/740. The `e2e-openstack` job won't succeed without it, but it does work with that change in our local testing.

**Comment by tomassedovic** [comment]
/retest

https://github.com/openshift/machine-config-operator/pull/740 is now merged and I just got an `Install complete` with no release image override locally.

File: pkg/types/openstack/validation/platform.go [hunk 8]
```diff
@@ -64,14 +67,29 @@ func ValidatePlatform(p *openstack.Platform, fldPath *field.Path, fetcher ValidV
...
+		expectedAPIVIP, _ := cidr.Host(&n.MachineCIDR.IPNet, 5)
+		if p.APIVIP != expectedAPIVIP.String() {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("apiVIP"), p.APIVIP, fmt.Errorf("the API VIP is expected to be %s", expectedAPIVIP).Error()))
+		}
 	}
 
 	if err := validate.IP(p.DNSVIP); err != nil {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("dnsVIP"), p.DNSVIP, err.Error()))
+	} else if n != nil {
+		expectedDNSVIP, _ := cidr.Host(&n.MachineCIDR.IPNet, 6)
```
Comment (abhinavdahiya) [inline pkg/types/openstack/validation/platform.go]: seems like https://github.com/openshift/installer/pull/2140 is making similar change.. in terms of adding validation..

File: pkg/types/openstack/validation/platform.go [hunk 9]
```diff
@@ -64,14 +67,29 @@ func ValidatePlatform(p *openstack.Platform, fldPath *field.Path, fetcher ValidV
 	// Validate VIP Values
 	if err := validate.IP(p.APIVIP); err != nil {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("apiVIP"), p.APIVIP, err.Error()))
-	}
-
-	if err := validate.IP(p.IngressVIP); err != nil {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("ingressVIP"), p.IngressVIP, err.Error()))
+	} else if n != nil {
+		expectedAPIVIP, _ := cidr.Host(&n.MachineCIDR.IPNet, 5)
```
Comment (abhinavdahiya) [inline pkg/types/openstack/validation/platform.go]: hmm, the user is not alllowed to set the IP?? Why do we even have this field in the install-config then???

Comment (abhinavdahiya) [reply pkg/types/openstack/validation/platform.go]: oops, the comment is mean for change on the next line...

**Comment by mandre** [comment]
This PR depends on https://github.com/openshift/release/pull/4629 for e2e-openstack job.

File: pkg/types/openstack/validation/platform.go (hunk 9 above)
Comment (tomassedovic) [reply pkg/types/openstack/validation/platform.go]: That's a good point, yeah. I think trown (who wrote this originally IIRC) thought that's the only way to pass the values throughout the installer, but we can actually just use functions that return the VIPs we need throughout.

Going to fix that.

Comment (tomassedovic) [reply pkg/types/openstack/validation/platform.go]: The top commit removes the VIPs from install-config:

https://github.com/openshift/installer/pull/1959/commits/6659b86d4b62f6f30f3d227f2b118e3135063e64

**Comment by tomassedovic** [comment]
/retest

I got `Install complete!` on this locally.

**Comment by tomassedovic** [comment]
Looks like our current CI issues were caused by: https://github.com/openshift/installer/pull/2086

We're investigating now. This should hopefully fix it:

https://github.com/openshift/machine-config-operator/pull/1050

**Comment by tomassedovic** [comment]
Okay, I've reproduced the issue locally (by rebasing this PR on top of master) and verified that https://github.com/openshift/machine-config-operator/pull/1050 fixes it.

**Comment by tomassedovic** [comment]
Rebased. The release image now has https://github.com/openshift/machine-config-operator/pull/1050 so this should pass e2e-openstack (unless I've messed the rebase up or there's another breakage).

**Comment by tomassedovic** [comment]
/hold cancel

The OpenStack job passed, I've validated it locally and this PR is rabased on top of master. I think we're ready to merge.

**Comment by Fedosin** [comment]
"/test e2e-aws-scaleup-rhel7" is broken, no need to retest it

**Pull request merged**